### PR TITLE
test: incorporates tests for short absence proof verification 

### PR DESCRIPTION
**Pull request by staheri14**
## Overview
After investigating the current implementation, verified that the nmt library is capable of verifying short absence proofs. This PR contains the necessary tests to demonstrate and test this capability.

In line with #210 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

**Review by cmwaters — DISMISSED** [review]
Really clearly written tests 👍

**Review by rootulp — APPROVED** [review]
+1 Callum's comment, really well written and clear tests. Great work!

File: proof_test.go [hunk 1]
```diff
@@ -707,3 +707,181 @@ func TestIsEmptyProofOverlapAbsenceProof(t *testing.T) {
...
+
+	//                                       Node_0_8                                  Tree Root
+	//                            /                            \
+	//                        /                                 \
+	//                  Node0_4                             Node4_8                  Non-Leaf Node
+	//               /            \                     /                \
+	//             /                \                 /                    \
+	//      Node_0_2            Node_2_4         Node4_6              Node6_8        Non-Leaf Node
+	//      /      \            /     \           /    \               /     \
+	// Node_0_1  Node_1_2  Node_2_3 Node_3_4  Node4_5  Node5_6  Node_6_7 Node_7_8    Leaf Hash
```
[the remark this answers is not in the document]
Comment (staheri14) [reply proof_test.go]: Great question! nodes are assigned indices based on the range of leaves they cover, with the upper bound being non-inclusive. For example, Node3_4 denotes a node that covers the 3rd leaf  (excluding the 4th leaf), while Node4_6 represents the node that covers the 4th and 5th leaves. I have clarified it in the code as well. https://github.com/celestiaorg/nmt/pull/217/commits/b763911192552410e6dcbe2d7666263bc47c0e9a

File: proof.go [hunk 2]
```diff
@@ -146,23 +146,22 @@ func (proof Proof) IsEmptyProof() bool {
...
-// empty. `data` items MUST be ordered according to their index in the tree,
-// with `data[0]` corresponding to the namespaced data at index `start`,
-//
-//	and the last element in `data` corresponding to the data item at index
-//	`end-1` of the tree.
+// which the namespace `proof` is generated. `leaves` contains the namespaced
+// leaves of the tree in the range of [`proof.start`, `proof.end`).
+// For an absence `proof`, the `leaves` is empty.
+// `leaves` items MUST be ordered according to their index in the tree,
+// with `data[0]` corresponding to the namespaced leaf at index `start`,
```
[the remark this answers is not in the document]
Comment (staheri14) [reply proof.go]: Fixed https://github.com/celestiaorg/nmt/pull/217/commits/5c67fe47f0e66f592acab664760bad48442a270c

File: proof.go [hunk 3]
```diff
@@ -146,23 +146,22 @@ func (proof Proof) IsEmptyProof() bool {
...
 // VerifyNamespace verifies a whole namespace, i.e. 1) it verifies inclusion of
-// the provided `data` in the tree (or the proof.leafHash in case of absence
-// proof) 2) it verifies that the namespace is complete i.e., the data items
-// matching the namespace ID `nID`  are within the range [`proof.start`,
-// `proof.end`) and no data of that namespace was left out. VerifyNamespace
-// deems an empty `proof` valid if the queried `nID` falls outside the namespace
-// range of the supplied `root` or if the `root` is empty
+// the provided `leaves` in the tree (or the proof.leafHash in case of
+// full/short absence proof) 2) it verifies that the namespace is complete
+// i.e., the data items matching the namespace ID `nID`  are within the range
```
[the remark this answers is not in the document]
Comment (staheri14) [reply proof.go]: I am going to open a PR to fix the usage of namespace ID in the nmt, however, there is no harm to remove this `ID` right now: https://github.com/celestiaorg/nmt/pull/217/commits/3b33555ecb8eeacafdb085541debd57efd36a2fe

File: proof_test.go [hunk 4]
```diff
@@ -716,16 +716,20 @@ func TestVerifyNamespace_ShortAbsenceProof_Valid(t *testing.T) {
...
+	//
 	//                                       Node_0_8                                  Tree Root
 	//                            /                            \
 	//                        /                                 \
-	//                  Node0_4                             Node4_8                  Non-Leaf Node
+	//                  Node0_4                             Node4_8                    Non-Leaf Node
 	//               /            \                     /                \
 	//             /                \                 /                    \
-	//      Node_0_2            Node_2_4         Node4_6              Node6_8        Non-Leaf Node
+	//      Node_0_2            Node_2_4         Node4_6              Node6_8          Non-Leaf Node
```
Comment (rootulp) [inline proof_test.go:730]: [optional] some contain a leading `_` and some don't. In my opinion, not a big deal

```suggestion
	//      Node0_2            Node2_4         Node4_6              Node6_8          Non-Leaf Node
```

File: proof_test.go [hunk 5]
```diff
@@ -716,16 +716,20 @@ func TestVerifyNamespace_ShortAbsenceProof_Valid(t *testing.T) {
 	qNS := []byte{5} // does not belong to the tree
 	root, err := tree.Root()
 	assert.NoError(t, err)
-
+	// In the following illustration, nodes are suffixed with the range
+	// of leaves they cover, with the upper bound being non-inclusive.
+	// For example, Node3_4 denotes a node that covers the 3rd leaf (excluding the 4th leaf),
+	// while Node4_6 represents the node that covers the 4th and 5th leaves.
+	//
 	//                                       Node_0_8                                  Tree Root
```
Comment (rootulp) [inline proof_test.go:724]:
```suggestion
	//                                       Node0_8                                  Tree Root
```

File: proof_test.go [hunk 6]
```diff
@@ -803,16 +807,20 @@ func TestVerifyNamespace_ShortAbsenceProof_Invalid(t *testing.T) {
 	qNS := []byte{7} // does not belong to the tree
 	root, err := tree.Root()
 	assert.NoError(t, err)
-
+	// In the following illustration, nodes are suffixed with the range
+	// of leaves they cover, with the upper bound being non-inclusive.
+	// For example, Node3_4 denotes a node that covers the 3rd leaf (excluding the 4th leaf),
+	// while Node4_6 represents the node that covers the 4th and 5th leaves.
+	//
 	//                                       Node_0_8                                  Tree Root
```
Comment (rootulp) [inline proof_test.go:815]:
```suggestion
	//                                       Node0_8                                  Tree Root
```

**Review by rootulp — DISMISSED** [review]
Still LGTM

File: proof_test.go (hunk 6 above)
Comment (staheri14) [reply proof_test.go]: Fixed https://github.com/celestiaorg/nmt/pull/217/commits/1e56c61cce1da94de94841728079635f45db847a

File: proof_test.go (hunk 5 above)
Comment (staheri14) [reply proof_test.go]: Fixed https://github.com/celestiaorg/nmt/pull/217/commits/1e56c61cce1da94de94841728079635f45db847a

File: proof_test.go (hunk 4 above)
Comment (staheri14) [reply proof_test.go]: Fixed https://github.com/celestiaorg/nmt/pull/217/commits/1e56c61cce1da94de94841728079635f45db847a

**Review by SweeXordious — APPROVED** [review]

**Pull request merged**